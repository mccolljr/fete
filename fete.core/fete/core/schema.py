"""Schema, the data model for everything."""

from typing import (
    Any,
    Callable,
    ClassVar,
    Dict,
    Generic,
    Iterator,
    List,
    Mapping,
    Optional,
    Tuple,
    Type,
    Union,
    TypeVar,
    overload,
)


from abc import ABC, abstractmethod
from datetime import datetime, timezone

# pylint: disable=invalid-name
_T_Any = TypeVar("_T_Any")
# pylint: enable=invalid-name


class SchemaDefinitionError(Exception):
    """An exception caused by an invalid schema definition."""

    def __init__(self, cls_name: str, problem: str):
        """Initialize a new SchemaDefinitionError."""
        self.cls_name = cls_name
        self.problem = problem
        super().__init__(f"{self.cls_name}: {self.problem}")


class _SchemaDef(Mapping[str, "Field[Any]"]):
    __slots__ = ("fields", "id_field")

    def __init__(self, *, id_field: str = None):
        self.fields: Dict[str, "Field[Any]"] = {}
        self.id_field: Optional[str] = id_field

    def __getitem__(self, __k: str) -> "Field[Any]":
        if __k not in self.fields:
            raise KeyError
        return self.fields[__k]

    def __setitem__(self, __k: str, val: "Field[Any]") -> None:
        if __k in self.fields:
            raise ValueError(f"field {repr(__k)} already defined")
        self.fields[__k] = val

    def __len__(self) -> int:
        return len(self.fields)

    def __iter__(self) -> Iterator[str]:
        return iter(self.fields)


class SchemaMeta(type):
    """The metaclass that all classes with a schema must inherit from."""

    __schema__: _SchemaDef

    def __new__(
        cls, name: str, bases: Tuple[type, ...], attrs: Dict[str, Any], **extra
    ):
        """Construct a new class with a schema."""
        attrs["__schema__"] = _SchemaDef(id_field=extra.pop("id", None))
        new_class = super().__new__(cls, name, bases, attrs)
        for base in new_class.__mro__:
            base_schema = getattr(base, "__schema__", None)
            if base_schema is not None:
                for k in base_schema:
                    if k not in new_class.__schema__:
                        new_class.__schema__[k] = base_schema[k]
        return new_class

    def __init__(
        cls, name: str, bases: Tuple[type, ...], attrs: Dict[str, Any], **_extra
    ):
        """Initialize and validate a newly created class with a schema."""
        super().__init__(name, bases, attrs)
        id_field = cls.__schema__.id_field
        if id_field is not None and id_field not in cls.__schema__:
            raise SchemaDefinitionError(
                cls.__name__, f"id field {repr(id_field)} does not exist"
            )


class SchemaBase(metaclass=SchemaMeta):
    """A base class for types that have a schema.

    This class provides utilities like a default initializing constructor,
    a string representation, and a to_dict method.
    """

    __schema__: ClassVar[_SchemaDef]

    def __init__(self, **kwargs):
        """Assign values to the object's fields based on the keyword arguments."""
        for field in self.__schema__.values():
            attr = field.attr_name
            name = field.field_name
            if name in kwargs:
                setattr(self, attr, kwargs[name])
            elif field.default is not None:
                default_val = field.default
                if callable(default_val):
                    default_val = default_val()
                setattr(self, attr, default_val)
            elif field.nullable:
                setattr(self, attr, None)
            else:
                raise ValueError(f"missing initializer for {name}")

    def to_dict(self) -> Dict[str, Any]:
        """Convert this object to its dictionary representation."""
        result = {}
        for field in self.__schema__.values():
            attr = field.attr_name
            name = field.field_name
            if hasattr(self, attr):
                result[name] = self.__to_dict_helper(field.kind, getattr(self, attr))
        return result

    def __to_dict_helper(self, kind: "FieldKind[Any]", val: Any) -> Any:
        if val is not None:
            if isinstance(kind, Object):
                assert isinstance(val, SchemaBase)
                return val.to_dict()

            if isinstance(kind, Collection):
                assert isinstance(val, list)
                return [self.__to_dict_helper(kind.of_kind, v) for v in val]

        return val

    def __str__(self):
        """Get a string representation of this object."""
        return f"<{self.__class__.__name__} {self.to_dict()}>"


class Field(Generic[_T_Any]):
    """A descriptor for a single field in a schema."""

    kind: "FieldKind[_T_Any]"
    default: Optional[Union[_T_Any, Callable[[], _T_Any]]]
    nullable: bool
    attr_name: str
    field_name: str

    def __init__(
        self,
        kind: Union["FieldKind[_T_Any]", Type["FieldKind[_T_Any]"]],
        *,
        name: Optional[str] = None,
        default: Optional[Union[_T_Any, Callable[[], _T_Any]]] = None,
        nullable: bool = True,
    ):
        """Initialize the Field.

        Keyword Args:
            name (str, optional): The field name. Defaults to None, meaning that the attr name will be used.
            default (value or callable, optional): Default value or value factory. Defaults to None.
            nullable (bool, optional): Is None a valid value for this field? Defaults to True.
        """
        self.kind = kind() if isinstance(kind, type) else kind
        self.default = default
        self.nullable = nullable
        if name is not None:
            self.field_name = name

    def __set_name__(self, owner: "SchemaMeta", name: str):
        """Save this field in the owner's schema."""
        self.attr_name = name
        if not hasattr(self, "_field_name"):
            self.field_name = name
        self._append_to_schema(owner)

    def _append_to_schema(self, owner: "SchemaMeta"):
        if not hasattr(owner, "__schema__"):
            raise RuntimeError(f"cannot register {self} to {owner}, no __schema__")
        if self.field_name in owner.__schema__:
            raise RuntimeError(f"duplicate definition for field {self.field_name}")
        owner.__schema__[self.field_name] = self

    @overload
    def __get__(self, obj: None, objtype: Any) -> "Field[_T_Any]":
        """Get the corresponding schema field from the owner."""
        ...

    @overload
    def __get__(self, obj: Any, objtype: Any) -> _T_Any:
        """Get the corresponding schema field value from the object."""
        ...

    def __get__(self, obj: Any, objtype=None) -> Union[_T_Any, "Field[_T_Any]"]:
        """Get the corresponding field or field value from the owner or object."""
        if obj is None:
            return self
        return obj.__dict__.get(f"__f_{self.attr_name}")

    def __set__(self, obj: Any, value: Any):
        """Set the corresponding field value on the object, respecting nullability and conversion."""
        if value is None and self.nullable:
            obj.__dict__[f"__f_{self.attr_name}"] = None
        else:
            obj.__dict__[f"__f_{self.attr_name}"] = self.kind.convert(value)

    def __delete__(self, obj: Any):
        """Remove the corresponding field value from the object."""
        data_name = f"__f_{self.attr_name}"
        if data_name in obj.__dict__:
            del obj.__dict__[data_name]


class FieldKind(Generic[_T_Any], ABC):
    """The base FieldKind type."""

    @abstractmethod
    def convert(self, value: Any) -> _T_Any:
        """Try to convert a value to the value type expected by this field.

        Args:
            value (Any): The value to try to convert

        Returns:
            The converted value

        Raises:
            An exception indicating why the conversion failed
        """

    @abstractmethod
    def validate(self, value: _T_Any):
        """Validate a value of this kind."""


class Int(FieldKind[int]):
    """A FieldType describing an int value."""

    def convert(self, value: Any) -> int:
        """Try to convert a value to an int."""
        if isinstance(value, str):
            return int(value)
        if isinstance(value, int):
            return value
        if isinstance(value, float):
            return int(value)
        raise ValueError(f"{repr(value)} cannot be converted to int")

    def validate(self, value: int):
        """Do nothing."""


class Float(FieldKind[float]):
    """A FieldType describing a float value."""

    def convert(self, value: Any) -> float:
        """Try to convert a value to a float."""
        if isinstance(value, str):
            return float(value)
        if isinstance(value, int):
            return float(value)
        if isinstance(value, float):
            return value
        raise ValueError(f"{repr(value)} cannot be converted to float")

    def validate(self, value: float):
        """Do nothing."""


class Str(FieldKind[str]):
    """A FieldType describing a str value."""

    def convert(self, value: Any) -> str:
        """Try to convert a value to a str."""
        if isinstance(value, str):
            return value
        if isinstance(value, bytes):
            return str(value, "utf-8")
        if isinstance(value, datetime):
            return value.isoformat()
        return str(value)

    def validate(self, value: str):
        """Do nothing."""


class Bool(FieldKind[bool]):
    """A FieldType describing a boolean value."""

    def convert(self, value: Any) -> bool:
        """Try to convert a value to a bool."""
        if isinstance(value, int):
            return value != 0
        if isinstance(value, bool):
            return value
        raise ValueError(f"{repr(value)} cannot be converted to bool")

    def validate(self, value: bool):
        """Do nothing."""


class Bytes(FieldKind[bytes]):
    """A FieldType describing a binary value."""

    def convert(self, value: Any) -> bytes:
        """Try to convert a value to bytes."""
        if isinstance(value, str):
            return bytes(value, "utf-8")
        if isinstance(value, bytes):
            return value
        raise ValueError(f"{repr(value)} cannot be converted to bytes")

    def validate(self, value: bytes):
        """Do nothing."""


class DateTime(FieldKind[datetime]):
    """A FieldType describing a datetime value."""

    def __init__(self, tzinfo: timezone = timezone.utc):
        """Create a new DateTime with the specified timezone (default UTC)."""
        self.tzinfo = tzinfo

    def convert(self, value: Any) -> datetime:
        """Try to convert a value to a datetime with the appropriate timezone."""
        if isinstance(value, str):
            return self._from_isofmt(value)
        if isinstance(value, datetime):
            return value
        raise ValueError(f"{repr(value)} cannot be converted to datetime")

    def _from_isofmt(self, src: str) -> datetime:
        try:
            return datetime.fromisoformat(src).astimezone(self.tzinfo)
        except ValueError:
            pass
        if src.endswith("Z"):
            return datetime.fromisoformat(f"{src[:-1]}+00:00")
        raise ValueError(f"{repr(src)} is not a valid ISO 8601 string")

    def validate(self, value: datetime):
        """Do nothing."""


class Collection(FieldKind[List[_T_Any]]):
    """A FieldType describing a list of some other FieldKind."""

    of_kind: FieldKind[_T_Any]

    def __init__(self, of_kind: Union[FieldKind[_T_Any], Type[FieldKind[_T_Any]]]):
        """Create a FieldKind to describe a list of values of some other FieldKind."""
        self.of_kind = of_kind if isinstance(of_kind, FieldKind) else of_kind()

    def convert(self, value: Any) -> List[_T_Any]:
        """Try to convert a value to a list of values of the appropriate FieldKind."""
        try:
            return [self.of_kind.convert(v) for v in iter(value)]
        except TypeError as terr:
            raise ValueError(
                f"{value} cannot be converted to list of {type(self.of_kind).__name__}"
            ) from terr

    def validate(self, value: List[_T_Any]):
        """Do nothing."""


TSchema = TypeVar("TSchema", bound=SchemaBase)


class Object(FieldKind[TSchema]):
    """A FieldType describing some object with its own schema."""

    of_typ: Type[TSchema]

    def __init__(self, of_typ: Type[TSchema]):
        """Create a FieldKind to describe an object with its own schema."""
        self.of_typ = of_typ

    def convert(self, value: Any) -> TSchema:
        """Attempt to convert a value to the appropriate object."""
        if isinstance(value, dict):
            return self.of_typ(**value)
        if isinstance(value, SchemaBase):
            return self.of_typ(**value.to_dict())
        raise ValueError(f"{value} cannot be converted to {type(self.of_typ).__name__}")

    def validate(self, value: TSchema):
        """Do nothing."""
