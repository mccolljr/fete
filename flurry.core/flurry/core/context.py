from typing import TYPE_CHECKING, Protocol, TypeVar

if TYPE_CHECKING:
    from .storage import Storage

# pylint: disable=invalid-name
_T_Storage = TypeVar("_T_Storage", bound="Storage")
# pylint: enable=invalid-name


class Context(Protocol[_T_Storage]):
    """The minimum interface that a Context type must provice."""

    storage: _T_Storage
