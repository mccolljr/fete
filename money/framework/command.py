from abc import ABCMeta, abstractmethod
from typing import Any, Dict, Generic, Optional, Tuple, TypeVar, Union

from money.framework import schema
from money.framework.storage import Storage


class CommandDefinitionError(Exception):
    """An exception caused by an invalid command class definition."""

    def __init__(self, query_name: str, problem: str):
        self.query_name = query_name
        self.problem = problem
        super().__init__(f"{self.query_name}: {self.problem}")


class CommandMeta(schema.SchemaMeta, ABCMeta):
    """The metaclass that all command classes must inherit from."""

    __cmd_mixin__: bool
    Result: Optional[schema.SchemaMeta]

    def __new__(cls, name: str, bases: Tuple[type, ...], attrs: Dict[str, Any]):
        attrs.setdefault("__cmd_mixin__", False)
        new_class = super().__new__(cls, name, bases, attrs)
        return new_class

    def __init__(cls, name: str, bases: Tuple[type, ...], attrs: Dict[str, Any]):
        super().__init__(name, bases, attrs)
        if not cls.__cmd_mixin__:
            CommandMeta._validate_exec(cls)
            CommandMeta._validate_result(cls)

    @staticmethod
    def _validate_result(the_cls: type):
        result_class = getattr(the_cls, "Result", None)
        if result_class is not None and not isinstance(result_class, schema.SchemaMeta):
            raise CommandDefinitionError(
                the_cls.__name__,
                f"Result must be a type (have {type(result_class)})",
            )

    @staticmethod
    def _validate_exec(the_cls: type):
        exec_method = getattr(the_cls, "exec", None)
        if exec_method is None or getattr(exec_method, "__isabstractmethod__", False):
            raise CommandDefinitionError(
                the_cls.__name__, "exec method must be defined"
            )
        if not callable(exec_method):
            raise CommandDefinitionError(the_cls.__name__, "exec must be callable")


TCommandResult = TypeVar("TCommandResult", bound=Union[schema.SchemaBase, None])


class CommandBase(Generic[TCommandResult], schema.SchemaBase, metaclass=CommandMeta):
    """The class that all commands must subclass.

    A subclass may specify a `Result` member which points to some class
    where `isinstance(Result, SchemaMeta)` is `True`, If `Result` is specified,
    this class will be used as the result of the query.

    If a subclass specifies a member `__cmd_mixin__` with a value of `True`,
    that class can serve as a common base class for multiple commands.
    """

    __cmd_mixin__ = True

    @abstractmethod
    async def exec(self, storage: Storage) -> TCommandResult:
        ...
