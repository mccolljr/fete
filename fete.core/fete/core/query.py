"""Queries, the Q in CQRS."""

from typing import Any, Dict, Generic, Tuple, TypeVar, Union
from abc import ABCMeta, abstractmethod

from . import schema
from .context import Context

# pylint: disable=invalid-name
_T_Context = TypeVar("_T_Context", bound=Context)
_T_Result = TypeVar("_T_Result", bound=Union[schema.SchemaBase, None])
# pylint: enable=invalid-name


class QueryDefinitionError(Exception):
    """An exception caused by an invalid query class definition."""

    def __init__(self, query_name: str, problem: str):
        """Initialize a new QueryDefinitionError."""
        self.query_name = query_name
        self.problem = problem
        super().__init__(f"{self.query_name}: {self.problem}")


class QueryMeta(ABCMeta, schema.SchemaMeta):
    """The metaclass all query classes must have."""

    __query_mixin__: bool
    Result: schema.SchemaMeta

    def __new__(
        cls, name: str, bases: Tuple[type, ...], attrs: Dict[str, Any], **extra
    ):
        """Construct a new Query class."""
        attrs.setdefault("__query_mixin__", extra.pop("mixin", False))
        mewcls = super().__new__(cls, name, bases, attrs)
        return mewcls

    def __init__(
        cls, name: str, bases: Tuple[type, ...], attrs: Dict[str, Any], **_extra
    ):
        """Initialize and validate a newly created Query class."""
        super().__init__(name, bases, attrs)
        if not cls.__query_mixin__:
            QueryMeta._validate_fetch(cls)
            QueryMeta._validate_result(cls)

    @staticmethod
    def _validate_result(the_cls: type):

        result_class = getattr(the_cls, "Result", None)
        if result_class is None:
            raise QueryDefinitionError(the_cls.__name__, "Result must be defined")
        if not isinstance(result_class, schema.SchemaMeta):
            raise QueryDefinitionError(
                the_cls.__name__,
                f"Result must be a type with a schema (have {type(result_class)})",
            )

    @staticmethod
    def _validate_fetch(the_cls: type):
        fetch = getattr(the_cls, "fetch", None)
        if fetch is None or getattr(fetch, "__isabstractmethod__", False):
            raise QueryDefinitionError(the_cls.__name__, "fetch method must be defined")
        if not callable(fetch):
            raise QueryDefinitionError(the_cls.__name__, "fetch must be callable")


class QueryBase(Generic[_T_Context, _T_Result], schema.SchemaBase, metaclass=QueryMeta):
    """The class that all queries must subclass.

    Each subclass must specify a `Result` member which points to some class
    where `isinstance(Result, SchemaMeta)` is `True`

    If a subclass specifies a member `__query_mixin__` with a value of `True`,
    that class can serve as a common base class for multiple queries.
    """

    __query_mixin__ = True

    @abstractmethod
    async def fetch(self, context: _T_Context) -> _T_Result:
        """Implement the fetch operation for this query."""
        ...
