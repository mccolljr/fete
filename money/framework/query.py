from __future__ import annotations
from abc import ABCMeta, abstractmethod

from typing import Any, Callable, Dict, Optional, Tuple, Union

import money.framework.schema as schema
from money.framework.storage import Storage


class QueryDefinitionError(Exception):
    def __init__(self, query_name: str, problem: str):
        self.query_name = query_name
        self.problem = problem
        super().__init__(f"{self.query_name}: {self.problem}")


class QueryMeta(ABCMeta):
    __query_mixin__: bool
    Result: schema.SchemaMeta
    Arguments: Optional[schema.SchemaMeta]
    fetch: Callable[[QueryBase, Any], Any]

    def __new__(cls, name: str, bases: Tuple[type, ...], attrs: Dict[str, Any]):
        attrs.setdefault("__query_mixin__", False)
        x = super().__new__(cls, name, bases, attrs)
        return x

    def __init__(cls, name: str, bases: Tuple[type, ...], attrs: Dict[str, Any]):
        if not cls.__query_mixin__:
            QueryMeta._validate_fetch(cls)
            QueryMeta._validate_result(cls)
            QueryMeta._validate_arguments(cls)

    @staticmethod
    def _validate_result(the_cls: type):
        Result = getattr(the_cls, "Result", None)
        if Result is None:
            raise QueryDefinitionError(the_cls.__name__, "Result must be defined")
        if not isinstance(Result, schema.SchemaMeta):
            raise QueryDefinitionError(
                the_cls.__name__, f"Result must be a type (have {type(Result)})"
            )

    @staticmethod
    def _validate_arguments(the_cls: type):
        Arguments = getattr(the_cls, "Arguments", None)
        if Arguments is None:
            return
        if not isinstance(Arguments, schema.SchemaMeta):
            raise QueryDefinitionError(
                the_cls.__name__, f"Arguments must be a type (have {type(Arguments)})"
            )

    @staticmethod
    def _validate_fetch(the_cls: type):
        fetch = getattr(the_cls, "fetch", None)
        if fetch is None or getattr(fetch, "__isabstractmethod__", False):
            raise QueryDefinitionError(the_cls.__name__, "fetch method must be defined")
        if not callable(fetch):
            raise QueryDefinitionError(the_cls.__name__, f"fetch must be callable")


class QueryBase(metaclass=QueryMeta):
    __query_mixin__ = True

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

    @abstractmethod
    async def fetch(self, storage: Storage) -> Union[schema.SchemaBase, None]:
        ...
