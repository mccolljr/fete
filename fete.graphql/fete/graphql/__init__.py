from .application import GraphqlApplication

__all__ = ["GraphqlApplication"]
