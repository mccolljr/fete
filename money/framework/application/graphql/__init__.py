from .generator import GraphqlGenerator
from .graphql import GraphqlApplication

__all__ = ["GraphqlGenerator", "GraphqlApplication"]
