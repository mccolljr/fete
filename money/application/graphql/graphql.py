"""Graphql over HTTP."""

import json
import logging
from typing import NamedTuple

import graphene
from aiohttp import web
from aiohttp_middlewares.cors import cors_middleware

from .generator import GraphqlGenerator
from ..application import Application
from money.storage import Storage


class CorsOptions(NamedTuple):
    """CORS configuration options."""

    allow_origin: str


class GraphqlApplication(Application):
    """An Application that provides a Graphql interface over HTTP."""

    def __init__(self, storage: Storage, cors_opts: CorsOptions = None):
        """Initialize the application."""
        super().__init__()
        self.storage = storage
        self.cors_opts = cors_opts

    @property
    def gql_schema(self) -> graphene.Schema:
        """Get the graphql schema serviced by this application."""
        existing = getattr(self, "__gql_schema", None)
        if existing is None:
            logging.info("building graphql schema")
            existing = GraphqlGenerator(self).generate_schema()
            setattr(self, "__gql_schema", existing)
            logging.info("schema built successfully")
        return existing

    def run(self, *, host: str = "localhost", port: int = 8080, **kwargs):
        """Run the server."""
        logging.info("===== LISTENING ON %s:%d =====", host, port)
        web.run_app(self._setup_app(), host=host, port=port, **kwargs)

    def _setup_app(self):
        _ = self.gql_schema
        web_app = web.Application(
            middlewares=[
                cors_middleware(origins=[self.cors_opts.allow_origin])
                if self.cors_opts is not None
                else cors_middleware(allow_all=True)
            ]
        )
        web_app.add_routes([web.post("/", self._handle_req)])
        return web_app

    async def _handle_req(self, req: web.Request):
        body = await req.json()
        query = body.get("query", None)
        var_vals = body.get("variables", None)
        result = await self.gql_schema.execute_async(
            query,
            variable_values=var_vals,
            context=graphene.Context(storage=self.storage),
        )
        if result.errors:
            return web.Response(
                content_type="application/json",
                body=json.dumps({"errors": ",".join(map(str, result.errors))}),
                status=500,
            )
        return web.Response(
            content_type="application/json", body=json.dumps(result.data), status=200
        )
