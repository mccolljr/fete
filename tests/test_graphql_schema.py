# pylint: disable=missing-class-docstring
from typing import cast

from flurry.core import schema
from flurry.graphql import GraphqlApplication
from flurry.core.query import QueryBase
from flurry.core.context import Context
from flurry.core.command import CommandBase

_APP = GraphqlApplication(cast(Context, None))


@_APP.query()
class FooQuery(QueryBase):
    a = schema.Field(schema.Str)
    b = schema.Field(schema.DateTime)

    class Result(schema.SchemaBase):
        x = schema.Field(schema.Float)
        y = schema.Field(schema.Collection(schema.Bool()))

    async def fetch(self, context: Context) -> Result:
        return self.Result(x=1.0, y=[True, False])


class QuuxData(schema.SchemaBase):
    q = schema.Field(schema.Str)
    r = schema.Field(schema.Str)
    s = schema.Field(schema.Float)


@_APP.command()
class BarCommand(CommandBase):
    a = schema.Field(schema.Str)
    b = schema.Field(schema.DateTime)
    z = schema.Field(schema.Object(QuuxData))

    class Result(schema.SchemaBase):
        x = schema.Field(schema.Float)
        y = schema.Field(schema.Object(QuuxData))

    async def exec(self, context: Context) -> Result:
        return self.Result(x=1, y=QuuxData())


def test_schema(snapshot):
    assert str(_APP.gql_schema) == snapshot(name="graphql-schema")
