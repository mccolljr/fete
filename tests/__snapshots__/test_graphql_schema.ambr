# name: test_schema[graphql-schema]
  '
  schema {
    query: QueryRoot
    mutation: MutationRoot
  }
  
  type QuuxData {
    q: String
    r: String
    s: Float
  }
  
  input QuuxDataInput {
    q: String = null
    r: String = null
    s: Float = null
  }
  
  type QueryRoot {
    FooQuery(a: String = null, b: DateTime = null): FooQueryResult
  }
  
  type FooQueryResult {
    x: Float
    y: [Boolean!]
  }
  
  """
  The `DateTime` scalar type represents a DateTime
  value as specified by
  [iso8601](https://en.wikipedia.org/wiki/ISO_8601).
  """
  scalar DateTime
  
  type MutationRoot {
    BarCommand(a: String = null, b: DateTime = null, z: QuuxDataInput = null): BarCommandResult
  }
  
  type BarCommandResult {
    x: Float
    y: QuuxData
  }
  
  '
---
